pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1053 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![allow(clippy::map_identity)]
+
+use cfg_traits::{
+	investments::{ForeignInvestment, Investment, InvestmentCollector},
+	SimpleCurrencyConversion, StatusNotificationHook, TokenSwaps,
+};
+use cfg_types::investments::{
+	CollectedAmount, ExecutedForeignCollectInvest, ExecutedForeignCollectRedeem,
+	ExecutedForeignDecrease, Swap,
+};
+use frame_support::{ensure, traits::Get, transactional};
+use sp_runtime::{
+	traits::{EnsureAdd, EnsureAddAssign, Zero},
+	DispatchError, DispatchResult,
+};
+
+use crate::{
+	errors::{InvestError, RedeemError},
+	types::{
+		InnerRedeemState, InvestState, InvestTransition, RedeemState, RedeemTransition,
+		TokenSwapReason,
+	},
+	CollectedRedemptionTrancheTokens, Config, Error, Event, ForeignInvestmentInfo,
+	ForeignInvestmentInfoOf, InvestmentState, Pallet, RedemptionState, SwapOf, TokenSwapOrderIds,
+};
+
+mod invest;
+mod redeem;
+
+// Hook execution for (partially) fulfilled token swaps which should be consumed
+// by `TokenSwaps`.
+impl<T: Config> StatusNotificationHook for Pallet<T> {
+	type Error = DispatchError;
+	type Id = T::TokenSwapOrderId;
+	type Status = SwapOf<T>;
+
+	fn notify_status_change(
+		id: T::TokenSwapOrderId,
+		status: SwapOf<T>,
+	) -> Result<(), DispatchError> {
+		let info = ForeignInvestmentInfo::<T>::get(id).ok_or(Error::<T>::InvestmentInfoNotFound)?;
+		let reason = info
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		match reason {
+			TokenSwapReason::Investment => {
+				let pre_state = InvestmentState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(InvestTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(InvestError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_invest_state_transition(&info.owner, info.id, post_state)
+			}
+			TokenSwapReason::Redemption => {
+				let pre_state = RedemptionState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(RedeemTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(RedeemError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_redeem_state_transition(&info.owner, info.id, post_state)
+			}
+		}
+	}
+}
+
+impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
+	type Amount = T::Balance;
+	type CollectInvestResult = ExecutedForeignCollectInvest<T::Balance>;
+	type CurrencyId = T::CurrencyId;
+	type Error = DispatchError;
+	type InvestmentId = T::InvestmentId;
+
+	#[transactional]
```

Comment:
Why is this `transactional` and other trait impls are not?

Reply:
The only trait impls missing `transactional` are the getters: `investment`, `redemption`, `accepted_payment_currency`, `accepted_payout_currency`.